lang/LangSource/PyrSlot.h

```diff
@@ -27,16 +27,398 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
+static constexpr uint64_t ptrTag = 0xFFFB000000000000; // boxed | 0x4003000000000000;
+
+static_assert(objHdrTag != intTag);
+static_assert(objHdrTag != symTag);
+static_assert(objHdrTag != charTag);
+static_assert(objHdrTag != ptrTag);
+
+static constexpr uint64_t nilValue = nilFalseTrueTag | 0x0000000000000000;
+static constexpr uint64_t falseValue = nilFalseTrueTag | 0x0000000000000001;
+static constexpr uint64_t trueValue = nilFalseTrueTag | 0x0000000000000002;
+
+template <typename T> struct PaddedNumber {
+    static_assert(sizeof(T) <= 4);
+    static_assert(std::is_arithmetic_v<T>);
+    static constexpr size_t numExtraBytes = 4 - sizeof(T);
+
+    constexpr explicit PaddedNumber(T v) noexcept: value(v) {}
+    constexpr PaddedNumber() = default;
+
+    ~PaddedNumber() noexcept = default;
+
+#if (BYTE_ORDER == LITTLE_ENDIAN) // TODO: this could be the wrong way around!
+    T value;
+    uint8_t paddingForValue[sizeof(uint32_t) + numExtraBytes] = { 0 };
+#else
+    uint8_t paddingForValue[sizeof(uint32_t) + numExtraBytes] = { 0 };
+    T value;
+#endif
+};
+
+static_assert(sizeof(PaddedNumber<uint32_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint16_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint8_t>) == 8);
+
+template <typename T> struct MaybePaddedPointer {
+    static_assert(std::is_pointer_v<T>);
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == BIG_ENDIAN)
+    int32_t padding;
+#endif
+
+    T ptr;
+
+    [[nodiscard]] T getPtr() const noexcept {
+#if POINTER_NEEDS_PADDING
+        return ptr; // If there is padding, the tag hasn't tainted the pointer, return it.
+#else
+        return reinterpret_cast<T>(reinterpret_cast<uintptr_t>(ptr) & pointerMask);
+#endif
+    }
+
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == LITTLE_ENDIAN)
+    int32_t padding;
+#endif
+};
+
+static_assert(sizeof(MaybePaddedPointer<void*>) == sizeof(double));
+
+// cpp reference
+template <class To, class From>
+std::enable_if_t<sizeof(To) == sizeof(From) && std::is_trivially_copyable_v<From> && std::is_trivially_copyable_v<To>,
+                 To>
+bit_cast(const From& src) noexcept {
+    static_assert(std::is_trivially_constructible_v<To>,
+                  "This implementation additionally requires "
+                  "destination type to be trivially constructible");
+    To dst;
+    std::memcpy(&dst, &src, sizeof(To));
+    return dst;
+}
+
+struct PyrNil {};
+
+union PyrSlot {
+private:
+    struct PrivateTag {};
+    constexpr PyrSlot(PrivateTag, uint64_t raw) noexcept: u_raw(raw) {}
+    constexpr PyrSlot(PrivateTag, double d) noexcept: u_double(d) {}
+
+    uint64_t u_raw;
+    double u_double;
+    PaddedNumber<char> u_char;
+    PaddedNumber<int32_t> u_int;
+    MaybePaddedPointer<void*> u_ptr;
+    MaybePaddedPointer<struct PyrObjectHdr*> u_objectHeader;
+    MaybePaddedPointer<struct PyrSymbol*> u_symbol;
+
+    template <uint64_t T> [[nodiscard]] bool tagChecker() const noexcept {
+        return isBoxed() && ((u_raw >> 48) == (T >> 48));
```

Comment:
Can we extract either the `48` literal or this entire `==` operation to something descriptive?